### PR TITLE
[CS2113T-W13-1] revisED

### PR DESCRIPTION
**Pull request by syncode98**
revisED helps students to take notes for the subjects that they are studying. It is optimised for CLI users so that the students can quickly add their notes for the subjects that they are studying,

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,34 +1,346 @@
...
+`NoFlashcardException` for the latter. If the topics have flashcards, then these are transferred to the 
+`SubjectQuiz` class, while the maximum score of the quiz is set to be the total number of flashcards in the list of flashcards
+present in the `SubjectQuiz` class . 
+
+The application now returns to the `startQuiz` method and the current score of the quiz is set to zero. Subsequently, the
+application begins printing the questions from the flashcards and checks the answer that you provide. If you want to stop the 
+quiz, use the `stop` command. The application will then print out the score that you obtain. If you complete the quiz
+then the application not only prints the score, but it also prints the incorrect answers from the quiz.
+
+The following diagram shows how you can initiate the quiz for a subject.
```
Comment (xingrong123) [inline docs/DeveloperGuide.md]: Should this be a solid arrow line instead? 

Capture1

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,34 +1,346 @@
...
+└── tasks.txt
+```
+
+This class also stores the data in a cascading manner, meaning that when a call to save topics is made, all the data 
+under the topics will be saved. Similarly, when a call to save subjects is made, all the data under the subjects
+including topics will be saved. The behavior mentioned is shown in the diagram below.
+
+![subject sq](https://user-images.githubusercontent.com/15065550/97132939-43c3eb80-1783-11eb-91c2-09b79eb2e2f5.png)
+![topic save sq](https://user-images.githubusercontent.com/15065550/97132990-74a42080-1783-11eb-91f8-f6143c0e38c4.png)
+
```
Comment (xingrong123) [inline docs/DeveloperGuide.md]: I like how this is sequence diagram is split into 2 so it is neater and easier to read.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,34 +1,346 @@
...
+`SubjectQuiz` class, while the maximum score of the quiz is set to be the total number of flashcards in the list of flashcards
+present in the `SubjectQuiz` class . 
+
+The application now returns to the `startQuiz` method and the current score of the quiz is set to zero. Subsequently, the
+application begins printing the questions from the flashcards and checks the answer that you provide. If you want to stop the 
+quiz, use the `stop` command. The application will then print out the score that you obtain. If you complete the quiz
+then the application not only prints the score, but it also prints the incorrect answers from the quiz.
+
+The following diagram shows how you can initiate the quiz for a subject.
+![first](https://user-images.githubusercontent.com/46095141/97313097-3866e200-18a1-11eb-9525-73e38ceb7cbe.png)
```
Comment (xingrong123) [inline docs/DeveloperGuide.md]: Should the method name be in quotation marks?

image

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,34 +1,346 @@
...
+├── results.json
+└── tasks.txt
+```
+
+This class also stores the data in a cascading manner, meaning that when a call to save topics is made, all the data 
+under the topics will be saved. Similarly, when a call to save subjects is made, all the data under the subjects
+including topics will be saved. The behavior mentioned is shown in the diagram below.
+
+![subject sq](https://user-images.githubusercontent.com/15065550/97132939-43c3eb80-1783-11eb-91c2-09b79eb2e2f5.png)
+![topic save sq](https://user-images.githubusercontent.com/15065550/97132990-74a42080-1783-11eb-91f8-f6143c0e38c4.png)
```
Comment (xingrong123) [inline docs/DeveloperGuide.md]: I think it would be good to standardize the way sequence diagrams in the document are drawn.

image

image

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,34 +1,346 @@
...
+segment to the next one. There are three segments within the application:
+* Subjects: This segment edits the list of all the subjects present within the application and takes care of relevant
+commands such as add subject, delete subject or find subject.
+* Subject: This segment is accessed through the `subject NAMEOFSUBJECT` command. The user would then look at the individual 
+subjects and then would be able to add either tasks or topics.
+* Topic : This segment is accessed through the `topic NAMEOFTOPIC` command. The user can add, delete or find flashcards
+in this segment.
+
+The code is broken down into different packages, which combine classes that perform a similar functionality. Here
+are the list of packages:
```
Comment (xingrong123) [inline docs/DeveloperGuide.md]: I think it would be good to add a class diagram to show the structure of the classes.

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,34 +1,346 @@
...
+including topics will be saved. The behavior mentioned is shown in the diagram below.
+
+![subject sq](https://user-images.githubusercontent.com/15065550/97132939-43c3eb80-1783-11eb-91c2-09b79eb2e2f5.png)
+![topic save sq](https://user-images.githubusercontent.com/15065550/97132990-74a42080-1783-11eb-91f8-f6143c0e38c4.png)
+
+##### Loading data
+
+The loading of data follows a similar cascading manner as saving, as can be seen in the diagram below.
 
+![subject load sq](https://user-images.githubusercontent.com/15065550/97138326-e59f0480-1792-11eb-8161-7653625d71c4.png)
```
Comment (xingrong123) [inline docs/DeveloperGuide.md]: Should there be a class label for this instance ?

image

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,34 +1,346 @@
...
+This class also stores the data in a cascading manner, meaning that when a call to save topics is made, all the data 
+under the topics will be saved. Similarly, when a call to save subjects is made, all the data under the subjects
+including topics will be saved. The behavior mentioned is shown in the diagram below.
+
+![subject sq](https://user-images.githubusercontent.com/15065550/97132939-43c3eb80-1783-11eb-91c2-09b79eb2e2f5.png)
+![topic save sq](https://user-images.githubusercontent.com/15065550/97132990-74a42080-1783-11eb-91f8-f6143c0e38c4.png)
+
+##### Loading data
+
+The loading of data follows a similar cascading manner as saving, as can be seen in the diagram below.
```
Comment (xingrong123) [inline docs/DeveloperGuide.md]: Should this be a solid arrow line instead?

image

**Review by xingrong123 — COMMENTED** [review]
Good job! A few mistakes in the sequence diagrams but it was very neat and easy to read.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,34 +1,346 @@
...
+        * Result
+        * ResultList
+        * TopicQuiz
+        * SubjectQuiz
+    - Flashcard
+    - Subject
+    - SubjectList
+    - Topic
+    - TopicList
+
```
Comment (LIU-YiFeng-1) [inline docs/DeveloperGuide.md]: Would it be better to have an overall design of the application for a clearer illustration?

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,34 +1,346 @@
...
+The `Storage` package holds the necessary classes for the quiz functionality of this application. The following are
+the classes in the package:an abstract`Quiz` class, a `Result` class, a `SubjectQuiz` class and a `TopicQuiz` class.
+
+#### Quiz class
+`Quiz` class is an abstract class in the `Quiz` package. It is holds the result of a quiz class and the list of flashcards
+from any subject or topic classes which the user calls the quiz for. Furthermore, it contains a checkAnswer() method the
+checks the answer that the user had given with the correct answer of the quiz. If the user enters the correct answer, the 
+existing score is incremented by one. Else, the contents of the flashcards and the incorrect answer provided by
+the user is transferred to the  `incorrectAnswers` list. Once the user finished the quiz, the application would print the 
+questions that the users did not answer correctly, along the the answer that was provided by the user.
```
Comment (LIU-YiFeng-1) [inline docs/DeveloperGuide.md]: would it be better to have a class diagram? I felt the explanation is a bit wordy.

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (flsqqq) [inline docs/DeveloperGuide.md]: Should the person using the app be referred to as "the user" instead of "you"?

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,34 +1,346 @@
...
+There are two classes inside this package, both of which are inside `Storage.java`. The following diagram shows the
+relationship between the two classes.
+
+![Storage uml s](https://user-images.githubusercontent.com/15065550/97129522-0eff6680-177a-11eb-85f5-41f4917bf8c3.png)
+
+#### StorageBuilder
+
+`StorageBuilder` class is a static inner class of the `Storage` class. It follows the builder pattern and its sole purpose
+is to create `Storage` instances. The [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) is used to deal 
+with the decrease in clarity due to the increase of the number parameters. Every instance of the `Storage` class should
```
Comment (LIU-YiFeng-1) [inline docs/DeveloperGuide.md:457]: Good use of external resources to guide interested readers to find out more.

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,34 +1,346 @@
...
+able to revise their subjects effectively through this application by adding various questions
+and answers in the form of flashcards. After that, students can refer to the flashcards to
+revise the subjects. Furthermore, they can take quizzes and see if they grasp the topics or subjects well
+enough to attend their school based exams.
+
+Students who find it hard to keep track of their assignments would also find this application
+useful since it lets the students add their tasks. The application then reminds the students
+regarding the upcoming tasks, which would prompt the students to prepare them.
+
+Students who find it difficult to organise woudl find this app useful since it
```
Comment (flsqqq) [inline docs/DeveloperGuide.md]: Should it be "would" instead?

**Review by LIU-YiFeng-1 — COMMENTED** [review]
Well Done! Easy to read document, I think it would be better if more class diagrams can be included.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,34 +1,346 @@
...
+### Testing the source code
+
+#### Setting up the project
+1. Download the source code from [here](https://github.com/AY2021S1-CS2113T-W13-1/tp/releases).
+2. The application is optimised for IntelliJ users. Download IntelliJ if you do not have it.
+3. Configure the Intellij for JDK 11, as described [here](https://se-education.org/guides/tutorials/intellijJdk.html).
+4. Import the project as a Gradle project, as described [here](https://se-education.org/guides/tutorials/intellijImportGradleProject.html).
+5. After the importing is complete, locate the `src/main/java/seedu.revised/Revised.java` file, right-click it, and choose
+`Run Revised.main()`.
+6. If the setup is correct, you should see soemthing like this:
```
Comment (flsqqq) [inline docs/DeveloperGuide.md]: Should it be "something"? There are occasional typos and grammar mistakes, perhaps they could be fixed to improve readability?

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (flsqqq) [inline docs/DeveloperGuide.md]: Should the diagrams be standardised (perhaps by using the same drawing software) to improve consistency?
image
image

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (chuhann) [inline docs/DeveloperGuide.md]: image

Consider changing "sd saving topic" to "alt", "opt", "par" or "loop" as "sd saving topic" is not a valid action

File: docs/DeveloperGuide.md (hunk 6 above)
Comment (chuhann) [inline docs/DeveloperGuide.md]: image

1. Consider changing "ref" to "alt", "opt", "par" or "loop" as "ref" is not a valid action.

2. Self-invocation to class should be a solid arrow instead of a dotted arrow for loadTopics(topicFolders), loadTasks(currentSubjectFolder) and loadResults(resultFile)

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (chuhann) [inline docs/DeveloperGuide.md]: image

1. The method name does not need quotation " ", around it

2. At the end of the lifecycle of "SubjectQuiz", consider removing ":Revised", ":SubjectParser", ":QuizSubjectCommand", ":SubjectQuiz", and ":UI" classes as the classes has already been called earlier. Consider replacing these with "X" to show the end of life line for each classes.

Comment (kaijiel24) [inline docs/DeveloperGuide.md]: image

I think you can omit the use of quotation marks (`""`) for your methods. For other diagrams you managed to do so. (see below)

image

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,34 +1,346 @@
...
+ refer to these to understand on how to use the application.
+ * Compatibility: The application runs on various operating systems: Windows, Linux etc
+ * Testability: The application is built using Gradle, a tool that offers various testing capabilities. Furthermore,
+ developers could also modify the input.txt file to provide a set of inputs and test if the application returns the 
+ expected output.
 
 ## Glossary
 
-* *glossary item* - Definition
+* *Revised* - The main class which initiates the whole application.
```
Comment (kaijiel24) [inline docs/DeveloperGuide.md]: Perhaps you can link the following items to their respective use case or vice versa (from their use case to glossary). This will help improve the navigability within you DG

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,34 +1,478 @@
...
+## Design & implementation <a name= "design"> </a>
 
-### Value proposition
+This section describes how RevisED has been designed and implemented so that developers can better understand the 
+designing of the code, with the help of visual aids.
 
-{Describe the value proposition: what problem does it solve?}
+The application has been broken down into different segments so that users would find it easy to navigate from one segment 
+to the next one. There are three segments within the application:
+* Main: This segment edits the list of all the subjects present within the application and takes care of relevant commands
```
Comment (max-wunan) [inline docs/DeveloperGuide.md]: It might be better for you guys to have a diagram for you overall architecture so that the readers can know which classes or interfaces link with/implement each other, instead of trying to understand the overall architecture through the individual sequence/class diagrams below

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,34 +1,478 @@
...
+`NoFlashcardException` for the latter. If the topics have flashcards, then these are transferred to the 
+`SubjectQuiz` class, while the maximum score of the quiz is set to be the total number of flashcards in the list of flashcards
+present in the `SubjectQuiz` class . 
+
+The application now returns to the `startQuiz` method and the current score of the quiz is set to zero. Subsequently, the
+application begins printing the questions from the flashcards and checks the answer that you provide. If you want to stop the 
+quiz, use the `stop` command. The application will then print out the score that you obtain. If you complete the quiz
+then the application not only prints the score, but it also prints the incorrect answers from the quiz.
+
+The following diagram shows how you can initiate the quiz for a subject.
```
Comment (max-wunan) [inline docs/DeveloperGuide.md]: It is a bit cramped here to have a bunch of diagrams with just a single line of explanation while there is no numbering or captions of all the diagrams. As a result it might be a bit confusing for readers. You guys can consider put captions/explanations, at least label with diagram is for what.
